### ISSUE TITLE
The time cost of the training process in step 1

### ISSUE THREAD

**Issue by Machine97** [cvg/SOLD2#20]
Hello, thank you for sharing!
In the training process of step 1, an epoch costs one to two hours, and you set up 200 epoch. Is this training time normal? Is there any way to improve the training speed?

**Comment by rpautrat** [comment]
Hi!

The given number of epochs 200 is more for the training on the Wireframe dataset (step 4). For synthetic training, less epochs are needed. 30 to 40 should be more than enough. It took a bit more than a day on a single GPU in our case. You can monitor the training metrics in Tensorboard and stop the training as soon as they stop increasing.

The training on the Wireframe dataset is however quite slow. It took a bit more than 4 days in our case to reach the 200 epochs. Fine-tuning the pre-trained model available on this repo will greatly accelerate this training time though.

**Comment by Machine97** [comment]
Thank you for your replying. In order to accelerate the training time, I used the pre-trained 'model sold2_synthetic.tar' you provided and changed the batch_size from 6 to 60 in the first step. Is that ok?

**Comment by rpautrat** [comment]
Sure, that should help making it faster and should not hinder the performance anyway.

**Comment by Machine97** [comment]
Hi,
In view of the slow training speed, I would like to ask if it is possible to change the code into multi-GPU training?

**Comment by rpautrat** [comment]
Hi, I am not sure I would have time to do this change in the future unfortunately... I recommend you to try it on your own if you really need it.

**Comment by Machine97** [comment]
> Hi, I am not sure I would have time to do this change in the future unfortunately... I recommend you to try it on your own if you really need it.

Ok, thank you very much! 
Another question, you recommend testing the results on a few samples of my dataset to check the quality of the output, and modifying the hyperparameters if need be. Could you please tell me how I test the results on a few samples of my dataset to check the quality of the output and where the hyperparameters in?

**Comment by rpautrat** [comment]
I meant checking the results visually by looking at a few sample detections and checking that the detections you want are there. I will also release later an evaluation pipeline to evaluate some metrics of the paper.

Regarding the hyperparameters, you mainly need to tune the detection parameters in 'line_detector_cfg' in `config/export_line_features.yaml`. I would try different values for `detect_thresh`, `inlier_thresh` (0.9 or 0.99 should work though), whether you want to use `candidate_suppression` or not (i.e. using NMS or not), and whether to use the `heatmap_refinement` or not. You can try with and without it, and keep the ones that seems the best for your application. The explanation of the first two parameters can be found in the original paper.

**Comment by Machine97** [comment]
But the result of step3 is .h5 file, like wireframe_train.h5 file, how can I get visualization result from the .h5 file and further check that the detections I want are there?

**Comment by rpautrat** [comment]
There is no ready-made code for this in this repo unfortunately. But it is just a matter of opening a .h5 file and visualizing the images in it. This can be done in just a few lines in Python.

**Comment by Machine97** [comment]
> There is no ready-made code for this in this repo unfortunately. But it is just a matter of opening a .h5 file and visualizing the images in it. This can be done in just a few lines in Python.

I tried as you said. For each ground truth, there are seven variables shown in the figure. How can I get the visualization image like the detection results in your paper in the help of these seven variables.
<img width="174" alt="截屏2021-08-03 下午5 38 52" src="https://user-images.githubusercontent.com/53359490/127995663-ea431df8-3baa-47e8-a28d-888a2eae8dc2.png">

**Comment by rpautrat** [comment]
Sorry, I should have given a bit more details. These seven variables should be coming from the output of step 2, no? These are only the raw outputs of homography adaptation. The two main variables are `heatmap_prob_mean` and `junc_prob_mean`, which are the averaged line and junction heatmaps.

But after running step 3, you should get a new .h5 file containing the line segments. You can then directly plot them with matplotlib, or using the following function: https://github.com/cvg/SOLD2/blob/71c3243f9d3a695788d0a6bfd134b9849425900a/misc/visualize_util.py#L340.

**Comment by Machine97** [comment]
> Sorry, I should have given a bit more details. These seven variables should be coming from the output of step 2, no? These are only the raw outputs of homography adaptation. The two main variables are `heatmap_prob_mean` and `junc_prob_mean`, which are the averaged line and junction heatmaps.
> 
> But after running step 3, you should get a new .h5 file containing the line segments. You can then directly plot them with matplotlib, or using the following function:
> 
> https://github.com/cvg/SOLD2/blob/71c3243f9d3a695788d0a6bfd134b9849425900a/misc/visualize_util.py#L340
> 
> .

The .h5 file obtained after running step 3 contain two variables including 'junctions' and 'line_map', the shape of these two variables are 300x2 and 300x300 respectively, but the shape of the input 'lines' in the function 'plot_lines' is 80x2x2, I am confused about this and I don't know how to plot the 'junctions' and 'line_map' by using the function 'plot_lines'.

**Comment by rpautrat** [comment]
Have a look at this part of the code: https://github.com/cvg/SOLD2/blob/71c3243f9d3a695788d0a6bfd134b9849425900a/model/line_detector.py#L108
There is an example on this line and the following ones showing how to get the line segments from junctions and line heatmap.

**Comment by Machine97** [comment]
Thank you so much, I have finished the pseudo ground truth visualization according to the method you said. And after the step4 training, I get the visualization results from the trained model on my own datasets. However, the detection results is so bad. The first image is the pseudo ground truth and the second one is detection result. To solve this problem, I observed that the train images in the same scene in wireframe dataset are continuously placed. I would like to ask if there are any requirements for the image placement order in the training set.
![29362147](https://user-images.githubusercontent.com/53359490/128174872-5c522119-a96e-418b-92cd-f49b38001bd0.jpg)
![37327027](https://user-images.githubusercontent.com/53359490/128174883-2ced63d0-06a4-4c6e-826e-f68786b268ba.jpg)
